src/core/mac/sub_mac.cpp

```diff
@@ -358,8 +358,23 @@ void SubMac::ProcessTransmitSecurity(void)
         mTransmitFrame.SetKeyId(mKeyId);
     }
 
-    VerifyOrExit(ShouldHandleTransmitSecurity());
-    VerifyOrExit(keyIdMode == Frame::kKeyIdMode1);
+#if OPENTHREAD_CONFIG_WAKEUP_COORDINATOR_ENABLE
+    if (mTransmitFrame.GetType() == Frame::kTypeMultipurpose)
+    {
+        // For now, process the Multipurpose frame security in SubMac even if the radio supports transmit security
+        // (which makes `ShouldHandleTransmitSecurity()` return false). The reason is that the Multipurpose frame
+        // support is a new thing in Thread and platforms that process transmit security for generic frame format in
+        // their radio layer, may not be ready to do the same for the Multipurpose frames.
+        //
+        // This behavior should be considered transitional and eventually removed in the future.
+        VerifyOrExit(keyIdMode == Frame::kKeyIdMode2);
+    }
+    else
+#endif
+    {
+        VerifyOrExit(ShouldHandleTransmitSecurity());
+        VerifyOrExit(keyIdMode == Frame::kKeyIdMode1);
+    }
```

Comment:
Yes, there should be one transmission at a time, which makes things simpler. However, if I recall correctly, the problem lies with the "enhanced acknowledgment" , which can also consume a security frame counter value. Enh-ACKs can be triggered between transmissions.

I recall there were discussions to change the Enh-ACK security frame model (so that it is tracked separately). I am not sure where we landed on this. Perhaps this may be already addressed by changes related to Enh-ACK and not an issue?

---

I can see both sides here:

- There is an advantage to (at least for now) performing the transmission security of multipurpose frames in `SubMac` to allow radio platforms to catch up. This may require a mechanism for transition (e.g., perhaps a new `OT_RADIO_CAP` to indicate that the radio is capable of performing this for multipurpose frames).
- Allowing this as a transient solution may result in it becoming the main approach used on some platform, which could lead to race conditions with handling the frame counter.

Reply:
How about if we add a new `otPlatRadioGetNextMacFrameCounter()` so we can ask the underlying radio platform to provide a frame counter for us to use?
* This would be used when `OT_RADIO_CAPS_TRANSMIT_SEC` is provided.
* We could add a new `OT_RADIO_CAPS_GET_NEXT_FRAME_COUNTER` to indicate that the radio supports this.
* We could then use it for multipurpose frames, so we can process transmission security in `SubMac`.
* This way, only one entity (the radio platform) would allocate and consume counter values.
* Also, we would not need to call `SetMacFrameCounter()` to synchronize the used counter with the radio platform later (as we get the counter from the radio directly).

What do you think?

Reply:
Ah, you're right. I didn't think about enh-ACKs.

Your proposal regarding `otPlatRadioGetNextMacFrameCounter` sounds interesting but I wonder if it's worth investing time in adding a new capability while in real life products it probably makes sense to simply support Multipurpose frames in the radio. In fact, I expected that for many such products we might even want to add `OT_RADIO_CAPS_WAKEUP_SEQ` to offload the entire wake-up sequence to the radio.

But I'll sleep on it and talk to my team about this :).

Reply:
Ok, we decided to revert this transitional code and simply assume platforms will handle MP frames in the radio.